Avoid File::Find during IDL generation

emw-apple opened the pull request:
#### eadc88e2217f19653e8c73db5ac16c3939fce399
<pre>
Avoid File::Find during IDL generation
<a href="https://bugs.webkit.org/show_bug.cgi?id=261364">https://bugs.webkit.org/show_bug.cgi?id=261364</a>
rdar://99146531

Reviewed by Alexey Proskuryakov.

Replace the include mechanism of generate-bindings.pl, which was
performing a recursive directory search of WebCore on every invocation,
with a lookup based on a precomputed list of IDL names.

The Xcode-based build generates this list already, in
IDLFileNamesList.txt. The CMake build does not have an equivalent list,
because it invokes the generator in multiple batches (once for WebCore,
WebCoreTestSupport, etc.). Change it to generate a file based on the
IDLs it&apos;s generating plus any additional IDLs which must be looked up
but not generated.

On my M2 Max machine, this is a 21% speedup of WebCore&apos;s installhdrs
time, from

  Time (mean ± σ):     88.534 s ±  8.544 s    [User: 6.732 s, System: 1.513 s]
  Range (min … max):   80.570 s … 100.172 s    10 runs

to

  Time (mean ± σ):     73.320 s ± 10.335 s    [User: 4.354 s, System: 1.363 s]
  Range (min … max):   66.565 s … 98.368 s    10 runs

* Source/WebCore/CMakeLists.txt: MockWebAuthenticationConfiguration is
  imported but not generated; add it to WebCoreTestSupport.
* Source/WebCore/DerivedSources.make: Recognize and pass
  --idlFileNamesList instead of --include arguments.
* Source/WebCore/WebCoreMacros.cmake: Generate the IDL file names list.
* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(new): Store path to IDL file names list.
(IDLFileForInterface): Read file names list instead of finding.
* Source/WebCore/bindings/scripts/generate-bindings-all.pl:
(buildDirectoryCache): Pass --idlFileNamesList argument.
* Source/WebCore/bindings/scripts/generate-bindings.pl:
(generateBindings): Ditto.
* Source/WebCore/testing/Internals.idl: Avoid having the IDL generator
  parse and look up attributes that require loading a bunch of
  additional IDLs. This prevents CMake builds from having to track all
  the Apple Pay IDLs in their WebCoreTestSupport, for instance.

* Tools/DumpRenderTree/DerivedSources-output.xcfilelist:
* Tools/DumpRenderTree/DerivedSources.make: Generate an
  IDLFileNamesList.txt and use it instead of includes.
* Tools/Scripts/webkitpy/bindings/main.py:
(BindingsTests.generate_from_idl):
(BindingsTests.generate_idl_file_names_list): Extracted logic, now
called twice: once to prepare a supplemental dependency file, once to
generate a list of all IDLs.
(BindingsTests):
(BindingsTests.generate_supplemental_dependency):
(BindingsTests.run_tests):
(BindingsTests.main):
* Tools/TestRunnerShared/CMakeLists.txt: Stop passing IDL_INCLUDES.
* Tools/WebKitTestRunner/CMakeLists.txt: Ditto.
* Tools/WebKitTestRunner/DerivedSources-output.xcfilelist:
* Tools/WebKitTestRunner/DerivedSources.make: Ditto DumpRenderTree
  changes.

Canonical link: <a href="https://commits.webkit.org/268561@main">https://commits.webkit.org/268561@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a841a0bf935c9e861a0fddd105376a4c300da35d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19803 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20227 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20845 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21696 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18504 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20038 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23487 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20372 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20052 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20023 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20014 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17225 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22551 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/19987 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17186 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18020 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24300 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18262 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18194 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22282 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18786 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15918 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17951 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4802 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22300 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18612 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebCore/testing/Internals.idl:
```diff
@@ -1057,7 +1057,9 @@ typedef (FetchRequest or FetchResponse) FetchObject;
     [Conditional=SERVICE_WORKER] Promise<undefined> terminateServiceWorker(ServiceWorker worker);
     [Conditional=SERVICE_WORKER] Promise<undefined> whenServiceWorkerIsTerminated(ServiceWorker worker);
 
+#if defined(ENABLE_APPLE_PAY) && ENABLE_APPLE_PAY
     [CallWith=CurrentDocument, Conditional=APPLE_PAY] readonly attribute MockPaymentCoordinator mockPaymentCoordinator;
+#endif
```
One thing that might be worth doing is making the code generator smart enough to understand that since this declaration is `Conditional=APPLE_PAY`, it does not need to be parsed unless `ENABLE_APPLE_PAY` is defined. Then, we don't need these redundant if guards, and we don't need to tell the build about IDLs that will generate fully ifdef'd-out code.

But that's a bit of a fundamental scope creep for the IDL generator, because it would be making decisions about feature definitions a priori, instead of just letting the preprocessor sort it out.

aproskuryakov replied:
What is the failure mode when one forgets to put these redundant #ifs? Given that this is an easy mistake to make, I'm hoping that it causes a build failure.

emw-apple replied:
Yeah, it's a failure when running the IDL generator, so should be pretty straightforward to diagnose.

emw-apple committed:
Avoid File::Find during IDL generation
https://bugs.webkit.org/show_bug.cgi?id=261364
rdar://99146531

Reviewed by Alexey Proskuryakov.

Replace the include mechanism of generate-bindings.pl, which was
performing a recursive directory search of WebCore on every invocation,
with a lookup based on a precomputed list of IDL names.

The Xcode-based build generates this list already, in
IDLFileNamesList.txt. The CMake build does not have an equivalent list,
because it invokes the generator in multiple batches (once for WebCore,
WebCoreTestSupport, etc.). Change it to generate a file based on the
IDLs it's generating plus any additional IDLs which must be looked up
but not generated.

On my M2 Max machine, this is a 21% speedup of WebCore's installhdrs
time, from

  Time (mean ± σ):     88.534 s ±  8.544 s    [User: 6.732 s, System: 1.513 s]
  Range (min … max):   80.570 s … 100.172 s    10 runs

to

  Time (mean ± σ):     73.320 s ± 10.335 s    [User: 4.354 s, System: 1.363 s]
  Range (min … max):   66.565 s … 98.368 s    10 runs

* Source/WebCore/CMakeLists.txt: MockWebAuthenticationConfiguration is
  imported but not generated; add it to WebCoreTestSupport.
* Source/WebCore/DerivedSources.make: Recognize and pass
  --idlFileNamesList instead of --include arguments.
* Source/WebCore/WebCoreMacros.cmake: Generate the IDL file names list.
* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(new): Store path to IDL file names list.
(IDLFileForInterface): Read file names list instead of finding.
* Source/WebCore/bindings/scripts/generate-bindings-all.pl:
(buildDirectoryCache): Pass --idlFileNamesList argument.
* Source/WebCore/bindings/scripts/generate-bindings.pl:
(generateBindings): Ditto.
* Source/WebCore/testing/Internals.idl: Avoid having the IDL generator
  parse and look up attributes that require loading a bunch of
  additional IDLs. This prevents CMake builds from having to track all
  the Apple Pay IDLs in their WebCoreTestSupport, for instance.

* Tools/DumpRenderTree/DerivedSources-output.xcfilelist:
* Tools/DumpRenderTree/DerivedSources.make: Generate an
  IDLFileNamesList.txt and use it instead of includes.
* Tools/Scripts/webkitpy/bindings/main.py:
(BindingsTests.generate_from_idl):
(BindingsTests.generate_idl_file_names_list): Extracted logic, now
called twice: once to prepare a supplemental dependency file, once to
generate a list of all IDLs.
(BindingsTests):
(BindingsTests.generate_supplemental_dependency):
(BindingsTests.run_tests):
(BindingsTests.main):
* Tools/TestRunnerShared/CMakeLists.txt: Stop passing IDL_INCLUDES.
* Tools/WebKitTestRunner/CMakeLists.txt: Ditto.
* Tools/WebKitTestRunner/DerivedSources-output.xcfilelist:
* Tools/WebKitTestRunner/DerivedSources.make: Ditto DumpRenderTree
  changes.

Canonical link: https://commits.webkit.org/268561@main
